### PR TITLE
Core: Initializing AdGuard's fork

### PR DESCRIPTION
**Pull request by ameshkov**
1. Applied AdGuard's patches to the sizzle.js
2. Added pre-processor task that compiles sizzle.adguard.js
(the version of Sizzle prepared to be used in ExtendedCss)

3. Fixed TokenSorter -- now we pass Sizzle tests
4. Fixed `nonnativeSelectorCache` -- it now works as intended
5. Fixed Sizzle's `speed` tests (added dependencies: domReady and text)

### CODE REVIEW

File: dist/sizzle.js [hunk 1]
```diff
@@ -1628,13 +1633,237 @@ function setFilters() {}
 setFilters.prototype = Expr.filters = Expr.pseudos;
 Expr.setFilters = new setFilters();
 
-tokenize = Sizzle.tokenize = function( selector, parseOnly ) {
+/**
+ * [AdGuard Patch]:
+ * Sorts the tokens in order to mitigate the performance issues caused by matching slow pseudos first:
+ * https://github.com/AdguardTeam/ExtendedCss/issues/55#issuecomment-364058745
+ */
+var TokenSorter = (function() {
```
Comment (theseanl) [inline dist/sizzle.js]: We don't need a "noun" here. We can make it simply return a `sort` function and save extra property lookup and code size.

File: dist/sizzle.js [hunk 2]
```diff
@@ -1628,13 +1633,237 @@ function setFilters() {}
...
+	}
+}
+
+/**
+ * [AdGuard Patch]:
+ * 
+ * This method processes parsed token groups, divides them into a number of selectors
+ * and makes sure that each selector's tokens are cached properly in Sizzle.
+ * 
+ * @typedef {Object} SelectorData
```
Comment (theseanl) [inline dist/sizzle.js]: This can be made more precise, otherwise what's the point of having a `typedef`?

File: dist/sizzle.js [hunk 3]
```diff
@@ -1628,13 +1633,237 @@ function setFilters() {}
...
+ * @property {string} selectorText A CSS selector text
+ * @property {Array} groups An array of token groups corresponding to that selector
+ * 
+ * @param {*} groups Token groups (see {@link Sizzle.tokenize})
+ * @returns {Array.<SelectorData>} An array of selectors data we got from the groups
+ */
+function tolerantTokenize(groups) {
```
Comment (theseanl) [inline dist/sizzle.js]: The purpose of this function is unclear from the function name. This need to be changed to something like `tokensToSelectors`.

File: dist/sizzle.js [hunk 4]
```diff
@@ -1687,12 +1916,41 @@ tokenize = Sizzle.tokenize = function( selector, parseOnly ) {
 	// Return the length of the invalid excess
 	// if we're just parsing
 	// Otherwise, throw an error or return tokens
-	return parseOnly ?
-		soFar.length :
-		soFar ?
-			Sizzle.error( selector ) :
-			// Cache the tokens
-			tokenCache( selector, groups ).slice( 0 );
+	var validLen = soFar.length;
```
Comment (theseanl) [inline dist/sizzle.js]: I think `invalidLen` would better suit as a variable name, because `soFar` denotes a selector fragment that couldn't be parsed.

**Review by theseanl — CHANGES_REQUESTED** [review]

File: dist/sizzle.js (hunk 2 above)
Comment (ameshkov) [reply dist/sizzle.js]: What do you mean? This is a precise type definition: two `@property` that follow `@typedef` describe this type.

File: dist/sizzle.js (hunk 3 above)
Comment (ameshkov) [reply dist/sizzle.js]: ++

File: dist/sizzle.js (hunk 4 above)
Comment (ameshkov) [reply dist/sizzle.js]: ++

File: dist/sizzle.js (hunk 1 above)
Comment (ameshkov) [reply dist/sizzle.js]: ++

File: dist/sizzle.js (hunk 2 above)
Comment (theseanl) [reply dist/sizzle.js]: I meant something like this: 
```js
/** @typedef {{ groups: Array<Array<?>>, selectorText: string }} SelectorData */
```

Comment (ameshkov) [reply dist/sizzle.js]: http://usejsdoc.org/tags-typedef.html

Comment (theseanl) [reply dist/sizzle.js]: Are you saying that the above is an invalid jsdoc?

Comment (ameshkov) [reply dist/sizzle.js]: Both are valid, mine is more precise.

Comment (theseanl) [reply dist/sizzle.js]: I don't think so. Please stay in point, what are you trying to say with a link to usejsdoc.org?

Comment (ameshkov) [reply dist/sizzle.js]: https://uploads.adguard.com/up04_o8u8r_sizzle-extcsssizzle.js_at_init__AdguardTeamsizzle-extcss.png

Comment (theseanl) [reply dist/sizzle.js]: ok, I was too accustomed to Closure-compiler style JSDoc :) It is meaningless in closure compiler and honestly I prefer it: https://groups.google.com/forum/#!topic/closure-compiler-discuss/vZNIh9vkJPU

Comment (ameshkov) [reply dist/sizzle.js]: I prefer describing properties as well and the closure-compiler's style does not let me do it.

Now we're both happy:
https://github.com/AdguardTeam/sizzle-extcss/pull/1/commits/57d2695f972d98377bf3667005490340c2fe7fc4

**Review by theseanl — APPROVED** [review]

**Pull request merged**